harness/determined/pytorch/_pytorch_trial.py

```diff
@@ -186,14 +185,38 @@ def _dataset_repro_warning(fn: str, data_obj: Any) -> str:
                 self.validation_loader = validation_data
 
     def run(self) -> None:
-        # We create the training_iterator here rather than in __init__ because we have to be careful
-        # to trigger its shutdown explicitly, to avoid hangs in when the user is using
-        # multiprocessing-based parallelism for their dataloader.
-        #
-        # We create it before loading state because we don't want the training_iterator shuffling
-        # values after we load state.
-        self.training_iterator = iter(self.training_loader)
-        try:
+        @contextlib.contextmanager
+        def defer(fn: Callable) -> Iterator[None]:
+            try:
+                yield
+            finally:
+                fn()
+
+        with contextlib.ExitStack() as exit_stack:
+            for callback in self.callbacks.values():
+                with self.prof.record_timing(
+                    f"callbacks.{callback.__class__.__name__}.on_trial_startup"
+                ):
+                    callback.on_trial_startup(self.latest_batch, self.env.latest_checkpoint)
+                    exit_stack.enter_context(defer(callback.on_trial_shutdown))
```

Comment:
this is cool

Reply:
indeed.

Reply:
One question, though. Are you deliberately _not_ timing the `on_trial_shutdown` callback? I see how it could be complex/expensive (you would probably create a new `Callable` that wraps `callback.on_trial_shutdown` into yet another `with self.prof.record_timing`.

Reply:
@mpkouznetsov no, it was an accident.

Shouldn't be expensive enough to matter.  Though the `enter_context(defer(callback.on_trial_shutdown))` shouldn't even be inside the initial `record_timing`.